File: backend/app/schemas/games_live.py
"""Live games — instructor live-play schemas (Live Games Phase 2C, 8-7).

Two shapes for a question on purpose: `LiveQuestionOut` (options carry
`is_correct`) is only ever returned from a staff-gated HTTP GET, never
broadcast — the WS run channel is shared by the instructor *and* every
student, so anything sent there uses `PublicQuestionOut` instead, which
strips `is_correct` entirely. Mixing these up would leak the answer key
to every connected student the instant a question starts.
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class PublicQuestionOptionOut(BaseModel):
    text: str


class PublicQuestionOut(BaseModel):
    """Broadcast shape — no `is_correct` anywhere. Sent over `question_started`."""
    id: UUID
    position: int
    prompt: str
    options: list[PublicQuestionOptionOut]
    time_limit_seconds: int
    max_points: int


class LiveQuestionOptionOut(BaseModel):
    text: str
    is_correct: bool


class LiveQuestionOut(BaseModel):
    """Staff-only shape — HTTP GET only, never broadcast."""
    id: UUID
    position: int
    prompt: str
    options: list[LiveQuestionOptionOut]
    time_limit_seconds: int
    max_points: int


class GameRunOut(BaseModel):
    id: UUID
    assignment_id: UUID
    run_no: int
    status: str
    current_question_position: int | None
    total_questions: int
    blackout_active: bool
    started_at: datetime | None
    ended_at: datetime | None


class RosterEntryOut(BaseModel):
    participant_id: UUID
    nickname: str
    avatar: str | None
    has_answered_current: bool


class LeaderboardEntryOut(BaseModel):
    participant_id: UUID
    nickname: str
    avatar: str | None
    score: int


class QuestionResultOut(BaseModel):
    index: int
    text: str
    is_correct: bool
    count: int
    pct: int


class RevealNameOut(BaseModel):
    participant_id: UUID
    real_name: str

File: frontend/src/api/games_live.ts
/** Live game instructor live-play API (Live Games Phase 2C, 8-7) — thin
 * wrapper over `/games/live/*`. Types mirror `schemas/games_live.py`. */
import { api } from "@/api/client";

export interface PublicQuestion {
  id: string;
  position: number;
  prompt: string;
  options: { text: string }[];
  time_limit_seconds: number;
  max_points: number;
}

export interface LiveQuestion {
  id: string;
  position: number;
  prompt: string;
  options: { text: string; is_correct: boolean }[];
  time_limit_seconds: number;
  max_points: number;
}

export interface GameRun {
  id: string;
  assignment_id: string;
  run_no: number;
  status: "lobby" | "live" | "ended";
  current_question_position: number | null;
  total_questions: number;
  blackout_active: boolean;
  started_at: string | null;
  ended_at: string | null;
}

export interface RosterEntry {
  participant_id: string;
  nickname: string;
  avatar: string | null;
  has_answered_current: boolean;
}

export interface LeaderboardEntry {
  participant_id: string;
  nickname: string;
  avatar: string | null;
  score: number;
}

export interface QuestionResult {
  index: number;
  text: string;
  is_correct: boolean;
  count: number;
  pct: number;
}

export const openRunApi = (assignmentId: string) =>
  api.post<GameRun>(`/games/live/assignments/${assignmentId}/runs`).then((r) => r.data);

export const getRunApi = (runId: string) =>
  api.get<GameRun>(`/games/live/runs/${runId}`).then((r) => r.data);

export const getCurrentQuestionApi = (runId: string) =>
  api.get<LiveQuestion>(`/games/live/runs/${runId}/question`).then((r) => r.data);

export const startRunApi = (runId: string) =>
  api.post<GameRun>(`/games/live/runs/${runId}/start`).then((r) => r.data);

export const revealRunApi = (runId: string) =>
  api.post<QuestionResult[]>(`/games/live/runs/${runId}/reveal`).then((r) => r.data);

export const nextQuestionApi = (runId: string) =>
  api.post<GameRun>(`/games/live/runs/${runId}/next`).then((r) => r.data);

export const restartRunApi = (runId: string) =>
  api.post<GameRun>(`/games/live/runs/${runId}/restart`).then((r) => r.data);

export const endRunApi = (runId: string) =>
  api.post<GameRun>(`/games/live/runs/${runId}/end`).then((r) => r.data);

export const getRosterApi = (runId: string) =>
  api.get<RosterEntry[]>(`/games/live/runs/${runId}/roster`).then((r) => r.data);

export const getLeaderboardApi = (runId: string) =>
  api.get<LeaderboardEntry[]>(`/games/live/runs/${runId}/leaderboard`).then((r) => r.data);

export const revealParticipantNameApi = (runId: string, participantId: string) =>
  api.get<{ participant_id: string; real_name: string }>(`/games/live/runs/${runId}/participants/${participantId}/reveal`).then((r) => r.data);

export const revealAllNamesApi = (runId: string) =>
  api.get<{ participant_id: string; real_name: string }[]>(`/games/live/runs/${runId}/reveal-all`).then((r) => r.data);
